Add Datadog metrics exporter

mx-psi opened the pull request:
**Description:** 
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->

- Add [Datadog](https://www.datadoghq.com/) metrics exporter using [DogStatsD](https://docs.datadoghq.com/developers/metrics/dogstatsd_metrics_submission) via the [`statsd`](https://pkg.go.dev/github.com/DataDog/datadog-go/statsd?tab=doc) package.

**Link to tracking Issue:** n/a

**Testing:** 

- Added unit tests.
- Built the Collector on this branch and performed manual tests with different receivers.

**Documentation:**

- The README briefly documents what the exporter does.
- The example configuration file documents every option possible values and usage.

mx-psi committed:
Create initial outline for Datadog exporter (#1)

* Add support for basic configuration options
* Documents configuration options

mx-psi committed:
Fix field name error

MetricsEndpoint was renamed to MetricsURL

mx-psi committed:
Address linter

mx-psi committed:
Bump collector version

mx-psi committed:
Initial DogStatsD implementation (#15)

Initial metrics exporter through DogStatsD with support for all metric types but summary and distribution

mx-psi committed:
Refactor configuration (#45)

* Refactor configuration
* Implement telemetry and tags configuration handling
* Update example configuration and README file
Co-authored-by: Kylian Serrania <kylian.serrania@datadoghq.com>

mx-psi committed:
Update dependencies and `make gofmt`
The collector was updated to 0.9.0 upstream

mx-psi committed:
Add support for summary and distribution metric types (#65)

* Add support for summary metric type

* Add support for distribution metrics

* Refactor metrics construction
- Drop name in Metrics (now they act as Metric values)
- Refactor constructor so that errors happen at compile-time

* Report Summary total sum and count values
Snapshot values are not filled in by OpenTelemetry

* Report p00 and p100 as `.min` and `.max`
This is more similar to what we do for our own non-additive type

* Keep hostname if it has not been overridden

mx-psi committed:
Disable Queue and Retry settings (#72)

These are handled by the statsd package.
OpenTelemetry docs are confusing and the default configuration (disabled)
is different from the one returned by "GetDefault..." functions

mx-psi committed:
Keep only DogStatsD parts
The API exporter will be added on a future PR

mx-psi committed:
Merge remote-tracking branch 'upstream/master' into dogstatsd

mx-psi committed:
Update core dependency and re-run style checks

mx-psi committed:
Add some more tests to make CodeCov happier

mx-psi committed:
Add final test for unspecified metrics

mx-psi commented:
This can be reviewed now: Codecov patch is failing but I think the patch coverage is reasonable (88.38%); some of the lines that Codecov indicates can't be tested (since we are not doing any configuration sanitization right now).

keitwb commented on exporter/datadogexporter/factory.go:
```diff
@@ -0,0 +1,109 @@
+}
+
+// CreateTracesExporter creates a traces exporter based on this config.
+func CreateTracesExporter(
```
You can omit this function and the corresponding WithTraces option above and it will by default consider traces unsupported.

mx-psi replied:
Makes sense, I removed it

mx-psi committed:
Address review comment
Remove unnecessary code on factory.go

keitwb commented on exporter/datadogexporter/config.go:
```diff
@@ -0,0 +1,117 @@
+	Endpoint string `mapstructure:"endpoint"`
+
+	// Telemetry states whether to send metrics
+	Telemetry bool `mapstructure:"telemetry"`
```
Is this like internal metrics about statsd or all metrics in general?  I would clarify the description of this.

mx-psi committed:
Clarify comment on telemetry configuration

mx-psi committed:
Further clarify comment

keitwb reviewed (approved):
LGTM

keitwb commented:
@mx-psi do try and improve coverage a bit.

mx-psi committed:
Improve test coverage further
- Remove configuration sanitization (currently unused)
- Remove metrics mapping error (currently unused)
- Add unit test for invalid configuration
- Add unit test for invalid exporter

mx-psi committed:
Merge remote-tracking branch 'upstream/master' into dogstatsd

mx-psi committed:
Remove usage of pdatautil package
Necessary to keep in line with other exporters after #953 was merged

albertvaka commented:
👋 Who else should review this before it can be merged? @asuresh4 (since you are assigned)?

asuresh4 commented:
> 👋 Who else should review this before it can be merged? @asuresh4 (since you are assigned)?

Did a quick pass through the PR on Friday and I think it looks good. I plan on taking a closer look today.

bogdandrutu commented:
@tylerbenson as one of the DataDog person can you review this? Or anyone else from DataDog can do it?

bogdandrutu commented on exporter/datadogexporter/factory.go:
```diff
@@ -0,0 +1,72 @@
+}
+
+// CreateMetricsExporter creates a metrics exporter based on this config.
+func CreateMetricsExporter(
```
This does not need to be public

tylerbenson commented:
I'll ask around to find someone that can review.

ericmustin commented:
@tylerbenson @mx-psi @bogdandrutu @albertvaka 

Hey friends. Just chiming in here, @mx-psi @tylerbenson and @albertvaka are all colleagues of mine at Datadog. I'm an approver on [opentelemetry-ruby](https://github.com/open-telemetry/opentelemetry-ruby) but am quite new to the collector, so I don't think I can add much to the code review, however generally I think this looks good to me.  I know that some of the metrics folks have been reviewing things over on [our fork](https://github.com/DataDog/opentelemetry-collector-contrib) and so I have a reasonable level of confidence here that Datadog as an org is both comfortable with this exporter and also plans to support it and add to it over time, should any issues arise. Additionally, I know there are a few limitations currently between what dogstatsd allows us to do and what our metric api / otel metrics permit, so while this PR might not have full coverage for every metric type,  I think Datadog plans to add additional (ie agentless) export options in the future, along with (ideally) exporters for traces at the collector level. I'll try to pop in on the SIG mtg later to say hello as I have a few other unrelated questions.

 I hope that helps! certainly please think of @mx-psi as the voice of reason here for all things datadog metrics going forward

bogdandrutu commented on exporter/datadogexporter/config.go:
```diff
@@ -0,0 +1,111 @@
+	// Endpoint is the DogStatsD address.
+	// The default value is 127.0.0.1:8125
+	// A Unix address is supported
+	Endpoint string `mapstructure:"endpoint"`
```
Can you use the https://github.com/open-telemetry/opentelemetry-collector/blob/master/config/confignet/confignet.go#L44 or https://github.com/open-telemetry/opentelemetry-collector/blob/master/config/confignet/confignet.go#L22?

mx-psi replied:
The `statsd` library picks up which protocol to use automatically (`udp` or `unixgram`) based on the `endpoint` field shape, so the `transport` field would have to be ignored in some cases (in particular, if `endpoint` starts with `unix://` it's considered UDS, and if certain environment variables are set it would be UDP). Would that behavior be okay you? Otherwise we would have to leave it as it is right now.

bogdandrutu reviewed (commented):
There are some reviews not addressed. Can you also split this based on https://github.com/open-telemetry/opentelemetry-collector/blob/master/CONTRIBUTING.md#when-adding-a-new-component

mx-psi commented:
@bogdandrutu 

> There are some reviews not addressed. Can you also split this based on https://github.com/open-telemetry/opentelemetry-collector/blob/master/CONTRIBUTING.md#when-adding-a-new-component

Sure! I will address the remaining comments next week and split it up into different PRs following the recently updated contributing guidelines. We plan on expanding the exporter's functionality so the configuration will not be final.

Is there any other recent change I should take into account for next week? I have tried to keep everything up to date with the rest of the exporters, but it's a bit hard to keep up with all the changes going on in the Collector.

> @/tylerbenson as one of the DataDog person can you review this? Or anyone else from DataDog can do it? 

As Eric stated in his message, this has already been reviewed by Datadog folks over at our fork; I will ask internally for people to chime in into the separate PRs next week.

mx-psi commented:
Hi all, after some internal discussion we have decided to rework this PR a bit so I am closing this. I will open new PRs in the future following the contribution guidelines with the reworked configuration and implementation. Thanks for the reviews, I will take them into account for the future PRs!
